File: app/src/main/java/com/cn29/aac/di/datasources/db/RoomModule.java
package com.cn29.aac.di.datasources.db;

import android.app.Application;

import com.cn29.aac.datasource.auth.db.AuthDao;
import com.cn29.aac.datasource.auth.db.AuthDb;
import com.cn29.aac.datasource.github.db.GithubDb;
import com.cn29.aac.datasource.github.db.RepoDao;
import com.cn29.aac.datasource.github.db.UserDao;
import com.cn29.aac.datasource.itunes.db.AlbumDao;
import com.cn29.aac.datasource.itunes.db.ArtistDao;
import com.cn29.aac.datasource.itunes.db.ItunesDb;

import javax.inject.Singleton;

import androidx.room.Room;
import dagger.Module;
import dagger.Provides;

/**
 * Created by Charles Ng on 16/10/2017.
 */

@Module
public class RoomModule {

  //Github Db instance and dao
  @Singleton
  @Provides
  GithubDb provideDb(Application app) {
    return Room.databaseBuilder(app, GithubDb.class, "github.db").build();
  }

  @Singleton
  @Provides
  UserDao provideUserDao(GithubDb db) {
    return db.userDao();
  }

  @Singleton
  @Provides
  RepoDao provideRepoDao(GithubDb db) {
    return db.repoDao();
  }


  //Auth Db instance and Dao
  @Singleton
  @Provides
  AuthDb provideAuthDb(Application application) {
    return Room.databaseBuilder(application, AuthDb.class, "auth.db").build();
  }

  @Singleton
  @Provides
  AuthDao provideAuthDao(AuthDb authDb) {
    return authDb.authDao();
  }


  //Itunes Db instance and Dao
  @Singleton
  @Provides
  ItunesDb provideItunesDb(Application application) {
    return Room.databaseBuilder(application, ItunesDb.class, "itunes.db").build();
  }

  @Singleton
  @Provides
  ArtistDao provideArtisDao(ItunesDb itunesDb) {
    return itunesDb.artistDao();
  }

  @Singleton
  @Provides
  AlbumDao provideAlbumDao(ItunesDb itunesDb) {
    return itunesDb.albumDao();
  }
}

File: app/src/main/java/com/cn29/aac/di/datasources/db/RoomModule.kt
package com.cn29.aac.di.datasources.db

import android.app.Application
import androidx.room.Room
import com.cn29.aac.datasource.auth.db.AuthDao
import com.cn29.aac.datasource.auth.db.AuthDb
import com.cn29.aac.datasource.github.db.GithubDb
import com.cn29.aac.datasource.github.db.RepoDao
import com.cn29.aac.datasource.github.db.UserDao
import com.cn29.aac.datasource.itunes.db.AlbumDao
import com.cn29.aac.datasource.itunes.db.ArtistDao
import com.cn29.aac.datasource.itunes.db.ItunesDb
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

/**
 * Created by Charles Ng on 16/10/2017.
 */
@Module
class RoomModule {
    //Github Db instance and dao
    @Singleton
    @Provides
    fun provideDb(app: Application?): GithubDb {
        return Room.databaseBuilder(app!!,
                                    GithubDb::class.java,
                                    "github.db").build()
    }

    @Singleton
    @Provides
    fun provideUserDao(db: GithubDb): UserDao {
        return db.userDao()
    }

    @Singleton
    @Provides
    fun provideRepoDao(db: GithubDb): RepoDao {
        return db.repoDao()
    }

    //Auth Db instance and Dao
    @Singleton
    @Provides
    fun provideAuthDb(application: Application?): AuthDb {
        return Room.databaseBuilder(application!!,
                                    AuthDb::class.java,
                                    "auth.db").build()
    }

    @Singleton
    @Provides
    fun provideAuthDao(authDb: AuthDb): AuthDao {
        return authDb.authDao()
    }

    //Itunes Db instance and Dao
    @Singleton
    @Provides
    fun provideItunesDb(application: Application?): ItunesDb {
        return Room.databaseBuilder(application!!,
                                    ItunesDb::class.java,
                                    "itunes.db").build()
    }

    @Singleton
    @Provides
    fun provideArtisDao(itunesDb: ItunesDb): ArtistDao {
        return itunesDb.artistDao()
    }

    @Singleton
    @Provides
    fun provideAlbumDao(itunesDb: ItunesDb): AlbumDao {
        return itunesDb.albumDao()
    }
}